src/update_halo.jl

```diff
@@ -709,8 +715,8 @@ function read_x2d!(gpurecvbuf::ROCDeviceArray{T}, A::ROCDeviceArray{T}, recvrang
 end
 
 # Write to the send buffer on the host from the array on the device (d2h).
-function write_d2h_async!(sendbuf::AbstractArray{T}, A::ROCArray{T}, sendranges::Array{UnitRange{T2},1}, signal::HSASignal) where T <: GGNumber where T2 <: Integer
-    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(sendbuf),sizeof(sendbuf),get_default_agent()))
+function write_d2h_async!(sendbuf::AbstractArray{T}, A::ROCArray{T}, sendranges::Array{UnitRange{T2},1}, signal::ROCSignal) where T <: GGNumber where T2 <: Integer
+    locked_ptr = convert(Ptr{T}, AMDGPU.Mem.lock(pointer(sendbuf),sizeof(sendbuf),AMDGPU.default_device()))
```

Comment:
This should probably use `AMDGPU.device(sendbuf)`, if `sendbuf <: ROCArray`.

Reply:
@omlins how should one implement the `sendbuf <: ROCArray` in Julian's suggestion to replace `AMDGPU.default_device()` by `AMDGPU.device(sendbuf)`?

Reply:
@jpsamaroo, `sendbuf` (and `recvbuf`) are actually no `ROCArray` and thus `AMDGPU.device(sendbuf)` fails. Either we leave it as such, or we may rather use `AMDGPU.device(A)` which is the corresponding `ROCArray` we want to copy to/from.

Reply:
Oh, sorry, I totally misread this! You want this to be host-locked, so you'd need `first(AMDGPU.devices(:cpu))`, probably.

Reply:
For host-locking, HSA will actually return a new pointer from the given device, thus I guess it should still be either `get_default_agent()` or `AMDGPU.device(A)`.

```
help?> AMDGPU.Mem.lock
  lock(ptr::Ptr, bytesize::Integer, device::ROCDevice)
  lock(ptr, bytesize)
  lock(a::Array, device)
  lock(a)

  Page-lock a host pointer allocated by the OS allocator and return a new pointer from the given device. For more information, see hsa_amd_memory_lock().

  See also: unlock

julia> 
```

Reply:
Any thoughts on ☝️ @jpsamaroo ?